src/Orleans/Serialization/OrleansJsonSerializer.cs

```diff
@@ -0,0 +1,179 @@
+                MissingMemberHandling = MissingMemberHandling.Ignore,
+                NullValueHandling = NullValueHandling.Ignore
+                //ConstructorHandling = ConstructorHandling.AllowNonPublicDefaultConstructor
+            };
+
+            settings.Converters.Add(new IPAddressConverter());
+            settings.Converters.Add(new IPEndPointConverter());
+            settings.Converters.Add(new GrainIdConverter());
+        }
+        
+        /// <summary>
+        /// Initializes the serializer
+        /// </summary>
+        /// <param name="logger">The logger to use to capture any serialization events</param>
+        public void Initialize(TraceLogger logger)
+        {
+            this.logger = logger;
+        }
+
+        /// <summary>
+        /// Creates a deep copy of an object
+        /// </summary>
+        /// <param name="source">The source object to be copy</param>
+        /// <returns>The copy that was created</returns>
+        public object DeepCopy(object source)
+        {
+            if (source == null)
+            {
+                return null;
+            }
+
+            var writer = new BinaryTokenStreamWriter();
+            Serialize(source, writer, source.GetType());
+            var retVal = Deserialize(source.GetType(), new BinaryTokenStreamReader(writer.ToByteArray()));
+            return retVal;
+        }
+
+        /// <summary>
+        /// Deserializes an object from a binary stream
+        /// </summary>
+        /// <param name="expectedType">The type that is expected to be deserialized</param>
+        /// <param name="reader">The <see cref="BinaryTokenStreamReader"/></param>
+        /// <returns>The deserialized object</returns>
+        public object Deserialize(Type expectedType, BinaryTokenStreamReader reader)
+        {
+            if (reader == null)
+            {
+                throw new ArgumentNullException("reader");
+            }
+
+            var str = (string)SerializationManager.DeserializeInner(typeof(string), reader);
+            return JsonConvert.DeserializeObject(str, expectedType, settings);
+        }
+
+        /// <summary>
+        /// Serializes an object to a binary stream
+        /// </summary>
+        /// <param name="item">The object to serialize</param>
+        /// <param name="writer">The <see cref="BinaryTokenStreamWriter"/></param>
+        /// <param name="expectedType">The type the deserializer should expect</param>
+        public void Serialize(object item, BinaryTokenStreamWriter writer, Type expectedType)
+        {
+            if (writer == null)
+            {
+                throw new ArgumentNullException("writer");
+            }
+
+            if (item == null)
+            {
+                writer.WriteNull();
+                return;
+            }
+
+            var str = JsonConvert.SerializeObject(item, expectedType, settings);
+            SerializationManager.SerializeInner(str, writer, typeof(string));
```

Comment:
I think you can call directly into BinaryTokenStreamWriter writer.Write(str). No need to loop back via`SerializationManager.SerializeInner`. Not 100% sure, but I think so. Same for deser.